contributing/building-a-bento-video-player.md

```diff
@@ -0,0 +1,627 @@
+
+On a host document, player components must dispatch the same events and implement the same methods so that playback and user interface can be coordinated successfully.
+
+**Preact components** can support this behavior by using a **`VideoWrapper`** that will use a specified `component`:
+
+```js
+return <VideoWrapper component="video" {...props} />
+```
+
+The `component` prop can be a string to specify a `<video>` element or an id reference to a component whose interface is similar to [`HTMLMediaElement`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement).
+
+However, most video players are embedded through an iframe so they should use **`VideoIframe`** instead. This is a specialized `VideoWrapper` that doesn't require an underlying `component`:
+
+```js
+return <VideoIframe {...props}>
+```
+
+To enable component support for **AMP documents**, our video player element must extend from a base class `VideoBaseElement`. This enables [actions](https://amp.dev/documentation/guides-and-tutorials/learn/amp-actions-and-events/#amp-video-and-other-video-elements_1) and [analytics](https://github.com/ampproject/amphtml/blob/main/extensions/amp-analytics/amp-video-analytics.md), and allows us to define further behavior specific to the AMP layer, like parsing element attributes into Preact props.
+
+This guide covers how to implement video player components that are internally implemented using these Preact and AMP components.
+
+## Getting Started
+
+Start by generating an extension specifying `--bento` and `--nojss`. We name our extension **`amp-fantastic-player`**, ending with `-player` according to our [guidelines for naming a third-party component](../spec/amp-3p-naming.md).
+
+```console
+amp make-extension --bento --nojss --name=amp-fantastic-player
+```
+
+## Directory Structure
+
+A [full directory for a Bento component](./building-a-bento-amp-extension.md#directory-structure) is generated, but this guide will cover the following files in particular:
+
+```
+/extensions/amp-fantastic-player/1.0/
+ ├── base-element.js               # Preact base element
+ ├── component.js                  # Preact implementation
+ ├── amp-my-fantastic-player.js    # Element's implementation
+ └── amp-my-fantastic-player.css   # Custom CSS
+```
+
+## Extend `VideoBaseElement` for AMP
+
+Our `BaseElement` should be a superclass of `VideoBaseElement`. In **`base-element.js`**, we change:
+
+```diff
+  import {MyFantasticPlayer} from './component';
```

Comment:
Suggestion for the sake of convention: consider putting the file names on the first line of these code blocks, a la
```
// base-element.js
import ...
```